Add state machine to drive migration states

harriswong opened the pull request:
## Description
This PR replaces the task queue with a state machine to drive the migration process. This PR will remember where the last "stopped" state is should an error occur. This allows the user to continue where they left off. For example, if the user installed the plugin but failed to activate it, the next time they run the migration, it will start from "activating" instead of the beginning. 

![image](https://github.com/Automattic/woocommerce-services/assets/572862/d7f36d8c-9710-4975-92ec-10e99008c04d)

### Related issue(s)
N/A

### Steps to test
#### Happy path - test the full migration
1. Go to `woocommerce-services/classes/class-wc-connect-service-settings-store.php` and replace the following `is_eligible_for_migration` function. This function decides if the modal should popup or not
```
		public function is_eligible_for_migration() {
			WC_Connect_Options::delete_option( 'wcshipping_migration_state' );
			$migration_state = WC_Connect_Options::get_option( 'wcshipping_migration_state' );

			$migration_pending = WC_Connect_WCST_To_WCShipping_Migration_State_Enum::COMPLETED !== $migration_state;
			$migration_enabled = $this->service_schemas_store->is_wcship_wctax_migration_enabled();

			return $migration_pending && $migration_enabled;
		}
```
2. Make sure you have "WooCommerce Shipping" and "WooCommerce Tax" in your plugins folder. We can't test the "install" part because they aren't part of dotorg yet.
3. Go to any order page with WCS&T enabled, click the WCS&T "Create shipping label" red button,
4. Click "Update now"
![image](https://github.com/Automattic/woocommerce-services/assets/572862/6d0b1591-9b4f-4ab4-8807-9a88ce91f769)
5. It should install, activate Woo Shipping & Woo Tax, deactivate WCS&T, and redirect you to the plugins page.
6. [Optional] Check DB's `wp_options` table, `wc_connect_options` should have the value `..."wcshipping_migration_state";i:12;}` appended at the end. Key thing here is `12` is the "done state". 

#### Test 404 on install and activate, fix the issue, then continue
1. Repeat Step 1 to 3 above. Refresh the page so it deletes the `wcshipping_migration_state` option.
2. Now, comment out `WC_Connect_Options::delete_option( 'wcshipping_migration_state' );` from the happy path step 1). In this test, we want to keep `wcshipping_migration_state` so we can test the exceptions.

What we want to test here is to mimic a 404 errors downloading from dotorg. Or some API errors during activation. Here, we will test by tripping the circuit breaker at the `wc-admin/plugins/install` and `wc-admin/plugins/activate` API calls. The goal is to test if the migration can continue where it last failed.

3. `npm run start`. Open up `client/components/migration/migration-runner.js`. Append `2` to the API URL so they fail and throw `404`.  Change the following
- `fetch( getBaseURL() + 'wc-admin/plugins/install', {` to `fetch( getBaseURL() + 'wc-admin/plugins/install2', {`
- `fetch( getBaseURL() + 'wc-admin/plugins/activate', {` to `fetch( getBaseURL() + 'wc-admin/plugins/activate2', {`
4. Click the "Update now" button
5. You should see this in the console because `/install2` failed 
![image](https://github.com/Automattic/woocommerce-services/assets/572862/29650ec2-27f8-42e8-8cc5-62270a2179b3)
The `migration_state` value should be `5` because it failed at installation. 
5.1 [optional] Check DB `wp_options`, `wc_connect_options` should have value appended at the end `"wcshipping_migration_state";i:5;}`.
6. Now, fix the issue by changing `/install2` back to `install`. Refresh the page. `install` should work, and activate will now fail. The installation should pick up from the "install" step.
7. Confirm it starts from "install" and failed at "activate". The `migration_state` should be 7. 
![image](https://github.com/Automattic/woocommerce-services/assets/572862/a0c7ae9f-4723-45b5-bfba-d139ab8c8e00)
7.1. [optional] Check DB `wp_options`, `wc_connect_options` should have value appended at the end `"wcshipping_migration_state";i:7;}`.
8. Now, fix the `activate2` to `activate`. The migration should pick up from state `7` and go through the whole migration.
![image](https://github.com/Automattic/woocommerce-services/assets/572862/2c4ccf52-1518-415a-9f98-0918ba2c2d1c)
9. [optional] Check DB `wp_options`, `wc_connect_options` should have value appended at the end `"wcshipping_migration_state";i:12;}`.
10. You should be on the plugins page `http://localhost/wp-admin/plugins.php?plugin_status=all&paged=1&s`. Confirm the WCS&T is deactivated, and Woo Shipping and Woo Tax are enabled.
![image](https://github.com/Automattic/woocommerce-services/assets/572862/5517130f-ba29-4afd-8e83-4157450fcd8b)


### Checklist

<!-- All testable code should have tests. -->
- [ ] unit tests

<!-- An entry in the changelog file is always required -->
- [ ] `changelog.txt` entry added
- [ ] `readme.txt` entry added

Commit:
Add state machine to drive migration states

Commit:
Rename state to better describe its name

Commit:
Refactor the API call out so the runNext function is decouple from API status check

Commit:
Update migration state enums to include all states in the machine

Commit:
Store the state in the wcshipping_migration_state option.

This allows us to remember where we were and we can always restart the
migration based on the last known state.

Commit:
Remove comment.

Commit:
Pass the wcshipping_migration_state option into the FeatureAnnouncement react component

Commit:
Fix a bug where restarting from an error state didn't proceed to the next state

Commit:
Return 304 if state is not modified instead of throwing 4xx.

Commit:
Fix 'wcshippingMigrationState' is already declared in the upper scope error

Commit:
Refactor the state machine logic into its own module

Commit:
Use wcs-client as path and re-enable redirection

Commit:
Added comments to the migration-runner module

harriswong commented on classes/class-wc-connect-account-settings.php:
```diff
@@ -39,21 +39,23 @@ public function get() {
+				'add_payment_method_url'     => $this->payment_methods_store->get_add_payment_method_url(),
+				'warnings'                   => array( 'payment_methods' => $payment_methods_warning ),
+				'is_eligible_to_migrate'     => $this->settings_store->is_eligible_for_migration(),
+				'wcshipping_migration_state' => $wcshipping_migration_state,
```
I considered having the React app to make an API `GET` call to retrieve this value from `useEffect()`. But we already pass `is_eligible_to_migrate` in, so it makes sense to me to pass `wcshipping_migration_state` in as well.

harriswong commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+        callback: stateErrorHandlingAPICall(MIGRATION_STATE_ENUM.ERROR_INSTALLING),
+    },
+    stateActivating: {
+        success: 'stateDeactivating', // TODO: This should be stateDBMigrating to migrate DB.
```
TODO: This should actually go to the state to migrate DB. This is going to be from another issue. Check https://github.com/woocommerce/shipping/issues/134#issuecomment-2191110937.

harriswong commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+    stateDB: {
+        success: 'stateDeactivating',
+        fail: 'stateErrorDB',
+        callback: new Promise((resolve) => resolve({status: 200})), // TODO: DB migration
```
TODO: We need an API to trigger the DB migration. Check https://github.com/woocommerce/shipping/issues/134#issuecomment-2192135701

Ferdev replied:
```suggestion
        callback: new Promise( ( resolve ) => resolve( { status: 200 } ) ), // TODO: DB migration
```

harriswong commented on woocommerce-services.php:
```diff
@@ -266,7 +266,7 @@ public static function plugin_deactivation() {
 
 			$migration_state = WC_Connect_Options::get_option( 'wcshipping_migration_state' );
-			if ( $migration_state === WC_Connect_WCST_To_WCShipping_Migration_State_Enum::STARTED ) {
+			if ( $migration_state !== WC_Connect_WCST_To_WCShipping_Migration_State_Enum::COMPLETED ) {
```
When the migration deactivates WCS&T, we won't be able to call the `connect/migration-flag` endpoint because the RESTful controller `WC_REST_Connect_Migration_Flag_Controller` is also deactivated. As a result, we rely on the `plugin_deactivation` hook to update the migration process.

If our migration state has started but not completed, and WCS&T is deactivated, then this means that the migration has successfully moved to the last step. We can safely update the migration flag to "completed".

Ferdev commented on client/components/migration/feature-announcement.jsx:
```diff
@@ -16,9 +15,11 @@ import bg from './images/wcshipping-migration.jpg';
 	isEligableToMigrate,
+	wcshippingMigrationState,
 } from 'woocommerce/woocommerce-services/state/shipping-label/selectors';
+import {installAndActivatePlugins} from './migration-runner';
```
```suggestion
import { installAndActivatePlugins } from './migration-runner';
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+ */
+const fetchAPICall = (apiFn) => async () => {
+    const apiResponse = await apiFn();
+    const apiJSONResponse = await apiResponse.json();
```
should we need to check `apiResponse` emptiness? Also, what does `json()` do if the response doesn't have a body?

harriswong replied:
Good questions. In `migration-runner.js`, it doesn't make use of the response as long as it didn't return an error code. 

I changed the code 74e0998c a bit so that
- `apiResponse` is always a Response object because it's returned by `fetch`. https://developer.mozilla.org/en-US/docs/Web/API/fetch#return_value. We shouldn't need to worry about emptiness here. 
- `.json()` is only called when it's > 400 status. The error message will be in the body.
- Return the `fetch` `Response` object instead, this `Response` object can be `null` or a stream. https://developer.mozilla.org/en-US/docs/Web/API/Response/body#value. `migration-runner.js` doesn't consume the response so it won't break the logic.

Let me know what you think!

Ferdev replied:
sounds perfect, thanks!

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+            return;
+        }
+
+        const currentMigrationState = migrationStateTransitions[migrationState];
```
should we check if `migrationState` is included in `migrationStateTransitions`?

harriswong replied:
Thanks! I added a function `getMigrationStateByName(stateName)` in 504f9c5f. It will now check if the state exists before using it. It will throw exception otherwise. Let me know what you think!

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+        }
+
+        const currentStateName = MIGRATION_ENUM_TO_STATE_NAME_MAP[previousMigrationState];
+        return migrationStateTransitions[currentStateName].success; // The next state is "success".
```
should we check if `currentStateName` exists in `migrationStateTransitions`?

harriswong replied:
_Same solution as https://github.com/Automattic/woocommerce-services/pull/2757#discussion_r1657757422_

Thanks! I added a function `getMigrationStateByName(stateName)` in 504f9c5f. It will now check if the state exists before using it. It will throw exception otherwise. Let me know what you think!

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+    let runAttemps = 0;
+    const maxAttempts = Object.keys(migrationStateTransitions).length; // The states don't loop, thus it can't be more than its size. Serve as a safe guard in case of infinite loop.
+    while ( runAttemps++ < maxAttempts ) {
+        nextMigrationStateToRun = await runNext(nextMigrationStateToRun);
```
I was gonna say we need a try/catch block here, but I just saw `runNext` can't throw an exception (it's very unlikely), so I think we're good :+1:

harriswong replied:
Yea! `runNext()` will follow the state diagram. Anything being thrown inside the "run" (ie. `callback()`) will get bubbled up into `runNext()`'s try-catch block. It will then go into the `fail` state. Any "failed" state will halt the runner and remember where it failed at.

Ferdev commented on client/components/migration/feature-announcement.jsx:
```diff
@@ -16,9 +15,11 @@ import bg from './images/wcshipping-migration.jpg';
+import {installAndActivatePlugins} from './migration-runner';
 
-const FeatureAnnouncement = ({ translate, isEligable }) => {
+const FeatureAnnouncement = ({ translate, isEligable, previousMigrationState }) => {
```
```suggestion
const FeatureAnnouncement = ( { translate, isEligable, previousMigrationState } ) => {
```

Ferdev commented on client/components/migration/feature-announcement.jsx:
```diff
@@ -31,78 +32,9 @@ const FeatureAnnouncement = ({ translate, isEligable }) => {
-		};
-
-		installAndActivatePlugins();
+		setIsUpdating(true);
+		installAndActivatePlugins(previousMigrationState);
+		setIsUpdating(false);
```
```suggestion
		setIsUpdating( true );
		installAndActivatePlugins( previousMigrationState );
		setIsUpdating( false );
```

Ferdev commented on client/components/migration/feature-announcement.jsx:
```diff
@@ -196,6 +128,7 @@ const FeatureAnnouncement = ({ translate, isEligable }) => {
 
 const mapStateToProps = (state, { siteId }) => ({
 	isEligable: isEligableToMigrate(state, siteId),
+	previousMigrationState: wcshippingMigrationState(state, siteId),
```
The existing code doesn't have proper formatting, but anyway... :smile: 

```suggestion
	previousMigrationState: wcshippingMigrationState( state, siteId ),
```

harriswong replied:
Hey @Ferdev! Thanks for catching these. How do you set up your `eslint` against this repo's rules? When I run `npm run eslint`, it doesn't say anything. This repo's eslintrc file https://github.com/Automattic/woocommerce-services/blob/trunk/.eslintrc.js also seems different.

I tried `npm install --save-dev @woocommerce/eslint-plugin` but that did not run properly in this repo.

I want to make sure we don't miss these eslint checks but the repo's precommit hooks don't seem to catch these at all. Thoughts?

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+ * @param {string} migrationState The numeric migration number from the PHP side. Check classes/class-wc-connect-wcst-to-wcshipping-migration-state-enum.php.
+ * @returns {Promise} Check fetch.
+ */
+const markMigrationStartedAPICall = (migrationState) => () =>
```
```suggestion
const markMigrationStartedAPICall = ( migrationState ) => () =>
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+        })
+    } );
+
+const stateErrorHandlingAPICall = (failedMigrationState) => () => {
```
```suggestion
const stateErrorHandlingAPICall = ( failedMigrationState ) => () => {
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+    // TODO: The error message doesn't show up anywhere. For now, update the flag in the option table so
+    // we know where it failed at.
+
+    return markMigrationStartedAPICall(failedMigrationState)();
```
```suggestion
    return markMigrationStartedAPICall( failedMigrationState )();
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+ * @param {function} apiFn The API function that uses fetch.
+ * @returns {Object} The API JSON response from fetch.
+ */
+const fetchAPICall = (apiFn) => async () => {
```
```suggestion
const fetchAPICall = ( apiFn ) => async () => {
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+const fetchAPICall = (apiFn) => async () => {
+    const apiResponse = await apiFn();
+    const apiJSONResponse = await apiResponse.json();
+    if (apiResponse.status >= 400) {
+        throw new Error(apiJSONResponse.message || translate("Failed to setup WooCommerce Shipping. Please try again."));
```
```suggestion
    if ( apiResponse.status >= 400 ) {
        throw new Error( apiJSONResponse.message || translate( "Failed to setup WooCommerce Shipping. Please try again." ) );
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+    stateInit: {
+        success: 'stateInstalling',
+        fail: 'stateInit',
+        callback: fetchAPICall(markMigrationStartedAPICall(MIGRATION_STATE_ENUM.STARTED)),
```
```suggestion
        callback: fetchAPICall( markMigrationStartedAPICall( MIGRATION_STATE_ENUM.STARTED ) ),
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+    stateErrorInit: {
+        success: 'stateInit',
+        fail: 'stateErrorInit',
+        callback: stateErrorHandlingAPICall(MIGRATION_STATE_ENUM.ERROR_STARTED),
```
```suggestion
        callback: stateErrorHandlingAPICall( MIGRATION_STATE_ENUM.ERROR_STARTED ),
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+    stateInstalling: {
+        success: 'stateActivating',
+        fail: 'stateErrorInstalling',
+        callback: fetchAPICall(installPluginAPICall),
```
```suggestion
        callback: fetchAPICall( installPluginAPICall ),
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+    stateErrorInstalling: {
+        success: 'stateInstalling',
+        fail: 'stateErrorInstalling',
+        callback: stateErrorHandlingAPICall(MIGRATION_STATE_ENUM.ERROR_INSTALLING),
```
```suggestion
        callback: stateErrorHandlingAPICall( MIGRATION_STATE_ENUM.ERROR_INSTALLING ),
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+    stateActivating: {
+        success: 'stateDeactivating', // TODO: This should be stateDBMigrating to migrate DB.
+        fail: 'stateErrorActivating',
+        callback: fetchAPICall(activatePluginAPICall),
```
```suggestion
        callback: fetchAPICall( activatePluginAPICall ),
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+    stateErrorActivating: {
+        success: 'stateActivating',
+        fail: 'stateErrorActivating',
+        callback: stateErrorHandlingAPICall(MIGRATION_STATE_ENUM.ERROR_ACTIVATING),
```
```suggestion
        callback: stateErrorHandlingAPICall( MIGRATION_STATE_ENUM.ERROR_ACTIVATING ),
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+    stateErrorDB: {
+        success: 'stateDBMigrating',
+        fail: 'stateErrorDB',
+        callback: stateErrorHandlingAPICall(MIGRATION_STATE_ENUM.ERROR_DB_MIGRATION),
```
```suggestion
        callback: stateErrorHandlingAPICall( MIGRATION_STATE_ENUM.ERROR_DB_MIGRATION ),
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+    stateDeactivating: {
+        success: 'stateDone',
+        fail: 'stateErrorDeactivating',
+        callback: fetchAPICall(deactivateWCSTPluginAPICall),
```
```suggestion
        callback: fetchAPICall( deactivateWCSTPluginAPICall ),
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+    stateErrorDeactivating: {
+        success: 'stateDeactivating',
+        fail: 'stateErrorDeactivating',
+        callback: stateErrorHandlingAPICall(MIGRATION_STATE_ENUM.ERROR_DEACTIVATING),
```
```suggestion
        callback: stateErrorHandlingAPICall( MIGRATION_STATE_ENUM.ERROR_DEACTIVATING ),
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+}
+
+// Any of the state that halts the machine. Including "done" or any errors.
+const stopStates = ['stateDone', 'stateErrorInstalling', 'stateErrorActivating', 'stateErrorDB', 'stateErrorDeactivating'];
```
```suggestion
const stopStates = [ 'stateDone', 'stateErrorInstalling', 'stateErrorActivating', 'stateErrorDB', 'stateErrorDeactivating' ];
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+ *
+ * @param {number} previousMigrationState This is the "wcshipping_migration_state" stored in the option table. This is the last state recorded for this migration before it stopped.
+ */
+const installAndActivatePlugins = async(previousMigrationState) => {
+    const runNext = async (migrationState) => {
+        if (!migrationState) {
```
```suggestion
const installAndActivatePlugins = async( previousMigrationState ) => {
    const runNext = async ( migrationState ) => {
        if ( !migrationState ) {
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+            return;
+        }
+
+        const currentMigrationState = migrationStateTransitions[migrationState];
```
```suggestion
        const currentMigrationState = migrationStateTransitions[ migrationState ];
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+     * @returns {string} The next state to run. The name of the state is the key in this object migrationStateTransitions.
+     */
+    const getNextStateToRun = () => {
+        if ( ! previousMigrationState) {
```
```suggestion
        if ( ! previousMigrationState ) {
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+    const getNextStateToRun = () => {
+        if ( ! previousMigrationState) {
+            // stateInit
+            return MIGRATION_ENUM_TO_STATE_NAME_MAP[2];
```
Not sure if numbers with brackets are supposed to be surrounded by spaces...

```suggestion
            return MIGRATION_ENUM_TO_STATE_NAME_MAP[ 2 ];
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+            return MIGRATION_ENUM_TO_STATE_NAME_MAP[2];
+        }
+
+        const currentStateName = MIGRATION_ENUM_TO_STATE_NAME_MAP[previousMigrationState];
```
```suggestion
        const currentStateName = MIGRATION_ENUM_TO_STATE_NAME_MAP[ previousMigrationState ];
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+        }
+
+        const currentStateName = MIGRATION_ENUM_TO_STATE_NAME_MAP[previousMigrationState];
+        return migrationStateTransitions[currentStateName].success; // The next state is "success".
```
```suggestion
        return migrationStateTransitions[ currentStateName ].success; // The next state is "success".
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+    // Run the migration chain from where it last stopped. If there is no record, start from the beginning.
+    let nextMigrationStateToRun = getNextStateToRun();
+    let runAttemps = 0;
+    const maxAttempts = Object.keys(migrationStateTransitions).length; // The states don't loop, thus it can't be more than its size. Serve as a safe guard in case of infinite loop.
```
```suggestion
    const maxAttempts = Object.keys( migrationStateTransitions ).length; // The states don't loop, thus it can't be more than its size. Serve as a safe guard in case of infinite loop.
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+    let runAttemps = 0;
+    const maxAttempts = Object.keys(migrationStateTransitions).length; // The states don't loop, thus it can't be more than its size. Serve as a safe guard in case of infinite loop.
+    while ( runAttemps++ < maxAttempts ) {
+        nextMigrationStateToRun = await runNext(nextMigrationStateToRun);
```
```suggestion
        nextMigrationStateToRun = await runNext( nextMigrationStateToRun );
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+    const maxAttempts = Object.keys(migrationStateTransitions).length; // The states don't loop, thus it can't be more than its size. Serve as a safe guard in case of infinite loop.
+    while ( runAttemps++ < maxAttempts ) {
+        nextMigrationStateToRun = await runNext(nextMigrationStateToRun);
+        if (stopStates.includes(nextMigrationStateToRun)) {
```
```suggestion
        if ( stopStates.includes( nextMigrationStateToRun ) ) {
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,251 @@
+        if (stopStates.includes(nextMigrationStateToRun)) {
+            // If this is the end of any of the error states, run the callback once more and then quit.
+            // This gives the state a chance to run its job before the machine quits.
+            await runNext(nextMigrationStateToRun);
```
```suggestion
            await runNext( nextMigrationStateToRun );
```

harriswong replied:
Hey @Ferdev Thanks for taking the time to spot all these out and make a suggestion for each. I hope that didn't take too much of your time 😢 We should probably look into fixing the eslint file for this repo. :/
Thank you again, appreciate it! It makes the "commit suggestion" and "add suggestion to batch" much easier!

Ferdev reviewed (changes requested):
Hey Harris! The implementation looks great, great job!!

I just left some comments mostly about formatting.

Commit:
Apply suggestions from code review

Lint fixes.

Co-authored-by: Fernando Espinosa <Ferdev@users.noreply.github.com>

Commit:
Wait for the migration to finish before setting isUpdating back to false

Commit:
Return the fetch Response object instead

Commit:
Added a function to check if state exists before using it.

Throw exception otherwise.

harriswong commented:
👋 Hi @Ferdev I think I addressed all the feedback on the PR, let me know what you think! Please feel free to merge for me if it looks good and if I am AFK.

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,260 @@
+            return;
+        }
+
+        const currentMigrationState = getMigrationStateByName(migrationState);
```
```suggestion
        const currentMigrationState = getMigrationStateByName( migrationState );
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,260 @@
+}
+
+// Before using the migration state, check if migration state exists, if not throw error.
+const getMigrationStateByName = (stateName) => {
+    if ( ! migrationStateTransitions[stateName] ) {
```
```suggestion
const getMigrationStateByName = ( stateName ) => {
    if ( ! migrationStateTransitions[ stateName ] ) {
```

Ferdev commented on client/components/migration/migration-runner.js:
```diff
@@ -0,0 +1,260 @@
+    if ( ! migrationStateTransitions[stateName] ) {
+        throw new Error( "Migration state [" + stateName + "] not found" );
+    }
+    return migrationStateTransitions[stateName];
```
```suggestion
    return migrationStateTransitions[ stateName ];
```

Ferdev committed:
Update client/components/migration/migration-runner.js

Ferdev committed:
Update client/components/migration/migration-runner.js

Ferdev committed:
Update client/components/migration/migration-runner.js

Ferdev reviewed (approved):
Thanks for the updates!! Added myself a couple of commits to fix a few more formatting issues I could find.

I think this is good to go. I just tested the full cycle and it worked without issues. Great job!! :rocket:
